src/Workspaces/Core/Portable/Utilities/BKTree.Builder.cs

```diff
@@ -0,0 +1,286 @@
+            //
+            //      Edge counts:
+            //      0   5560
+            //      1   1884
+            //      2   887
+            //      3   527
+            //      4   322
+            //      5   200
+            //      6   114
+            //      7   69
+            //      8   47
+            //      9   20
+            //      10  8
+            //      11  10
+            //      12  7
+            //      13  4
+            //      15  1
+            //      16  1
+            //      54  1
+            //
+            //
+            // i.e. The number of nodes with edge-counts less than or equal to four is: 5560+1884+887+527+322=9180.
+            // This is 95% of the total number of edges we are adding.  Looking at many other dlls
+            // we found that this ratio stays true across the board.  i.e. with all dlls, 95% of nodes
+            // have 4 or less edges.
+            //
+            // So, to optimize things, we pre-alloc a single array with space for 4 edges for each 
+            // node we're going to add.  Each node then gets that much space to store edge information.
+            // If it needs more than that space, then we have a fall-over dictionary that it can store 
+            // information in.
+            // 
+            // Once building is complete, the GC only needs to deallocate this single array and the
+            // spillover dictionaries.
+            //
+            // This approach produces 1/20th the amount of garbage while building the tree.
+            //
+            // Each node at index i has its edges in this array in the range [4*i, 4*i + 4);
+            private readonly Edge[] _compactEdges;
+            private readonly BuilderNode[] _builderNodes;
+
+            public Builder(IEnumerable<string> values)
+            {
+                var distinctValues = values.Where(v => v.Length > 0).Distinct(CaseInsensitiveComparison.Comparer).ToArray();
+                var charCount = Enumerable.Sum<string>(values,(Func<string, int>)(v => (int)v.Length));
+
+                _allLowerCaseCharacters = new char[charCount];
+                _characterSpans = new TextSpan[distinctValues.Length];
+
+                var characterIndex = 0;
+                for (int i = 0; i < distinctValues.Length; i++)
+                {
+                    var value = distinctValues[i];
+                    _characterSpans[i] = new TextSpan(characterIndex, value.Length);
+
+                    foreach (var ch in value)
+                    {
+                        _allLowerCaseCharacters[characterIndex] = char.ToLower(ch);
+                        characterIndex++;
+                    }
+                }
+
+                // We will have one node for each string value that we are adding.
+                _builderNodes = new BuilderNode[distinctValues.Length];
+                _compactEdges = new Edge[distinctValues.Length * CompactEdgeAllocationSize];
```

Comment:
I think this will scale better honestly.  The alternative implementation would be to have dictionaries for each item.  That's going to be 40k dictionaries being created, putting tons of strain on the GC.

> to figure out the sum of the lengths of those 38k names if you'd like that data

It's not hugely relevant IMO.  Right now we're *already* spending at least that much in SymbolTreeInfo** (as it has a string allocated for each of those unique names).  This approach at least puts far less pressure on the system by attempting to only allocate and use a few objects, while still representing a very large space of values.

** And i'd really like to go back and clean up SymbolTreeInfo to also not be string based. However, it's complicated as we need to call string comparison functions, and i'm not sure if those same functions exist for char[]s.